src/Repository/DeviceCodeRepository.php

```diff
@@ -0,0 +1,190 @@
+    }
+
+    public function getNewDeviceCode(): DeviceCodeEntityInterface
+    {
+        return new DeviceCodeEntity();
+    }
+
+    public function persistDeviceCode(DeviceCodeEntityInterface $deviceCodeEntity): void
+    {
+        $deviceCode = $this->deviceCodeManager->find($deviceCodeEntity->getIdentifier());
+
+        if (null === $deviceCode) {
+            $deviceCode = $this->buildDeviceCodeModel($deviceCodeEntity);
+        } else {
+            $deviceCode = $this->updateDeviceCodeModel($deviceCode, $deviceCodeEntity);
+        }
+
+        $this->deviceCodeManager->save($deviceCode);
+    }
+
+    public function approveDeviceCode(string $userCode, string $userId): void
+    {
+        $deviceCode = $this->deviceCodeManager->findByUserCode($userCode);
+
+        if (false === $deviceCode instanceof DeviceCodeInterface) {
+            throw OAuthServerException::invalidRequest('device_code', 'Device code does not exist');
+        }
+
+        if ($deviceCode->isRevoked()) {
+            throw OAuthServerException::invalidRequest('device_code', 'Device code has been revoked');
+        }
+
+        if ('' === $userId) {
+            throw OAuthServerException::invalidRequest('user_id', 'User ID is required');
+        }
+
+        $deviceCode->setUserIdentifier($userId);
+        $deviceCode->setUserApproved(true);
+
+        $this->deviceCodeManager->save($deviceCode);
+    }
+
+    public function getDeviceCodeEntityByDeviceCode(string $deviceCodeEntity): ?DeviceCodeEntityInterface
+    {
+        $deviceCode = $this->deviceCodeManager->find($deviceCodeEntity);
+
+        if (null === $deviceCode) {
+            return null;
+        }
+
+        return $this->buildDeviceCodeEntity($deviceCode);
+    }
+
+    public function revokeDeviceCode(string $codeId): void
+    {
+        $deviceCode = $this->deviceCodeManager->find($codeId);
+
+        if (null === $deviceCode) {
+            return;
+        }
+
+        $deviceCode->revoke();
+
+        $this->deviceCodeManager->save($deviceCode);
+    }
+
+    public function isDeviceCodeRevoked(string $codeId): bool
+    {
+        $deviceCode = $this->deviceCodeManager->find($codeId);
+
+        if (null === $deviceCode) {
+            return true;
+        }
+
+        return $deviceCode->isRevoked();
+    }
+
+    private function buildDeviceCodeEntity(DeviceCodeInterface $deviceCode): DeviceCodeEntity
+    {
+        $deviceCodeEntity = new DeviceCodeEntity();
+        $deviceCodeEntity->setIdentifier($deviceCode->getIdentifier());
+        $deviceCodeEntity->setExpiryDateTime($deviceCode->getExpiry());
+        $client = $this->clientRepository->getClientEntity($deviceCode->getClient()->getIdentifier());
+        if ($client) {
+            $deviceCodeEntity->setClient($client);
+        }
+        if ($deviceCode->getUserIdentifier()) {
+            $deviceCodeEntity->setUserIdentifier($deviceCode->getUserIdentifier());
+        }
```

Comment:
Prevent the casting of a string to a boolean value, and prevent calling the same method twice
```suggestion
        if (null !== $userIdentifier = $deviceCode->getUserIdentifier()) {
            $deviceCodeEntity->setUserIdentifier($userIdentifier);
        }
```